### PR TITLE
egressip: fix duplicate IP assignment on control-plane restart

### PR DESCRIPTION
**Pull request by cgoncalves**
When the ovn-k control-plane pods restart, two EgressIP CRs with the same IP in their specs could both get the IP assigned to nodes if one already had the IP assigned in status and the other didn't.

```console
$ kubectl get eip
NAME                     EGRESSIPS        ASSIGNED NODE                           ASSIGNED EGRESSIPS
egressip-dual-vlan94-1   192.168.118.30   cnfdc6.t5g-dev.eng.rdu2.dc.redhat.com   192.168.118.30
egressip-dual-vlan94-2   192.168.118.30                                           

$ kubectl -n openshift-ovn-kubernetes delete pods -l app=ovnkube-control-plane
pod "ovnkube-control-plane-6579d7875f-6v6zc" deleted
pod "ovnkube-control-plane-6579d7875f-7nggg" deleted

$ kubectl get eip
NAME                     EGRESSIPS        ASSIGNED NODE                           ASSIGNED EGRESSIPS
egressip-dual-vlan94-1   192.168.118.30   cnfdc6.t5g-dev.eng.rdu2.dc.redhat.com   192.168.118.30
egressip-dual-vlan94-2   192.168.118.30   cnfdc6.t5g-dev.eng.rdu2.dc.redhat.com   192.168.118.30
```

When WatchEgressNodes() processes node ADD events and calls addEgressNode(), it triggers reconcileEgressIP() for unassigned EgressIPs. At this point, the node allocator cache was empty and hadn't been populated with existing EgressIP assignments from their statuses. This allowed an unassigned EgressIP to get a duplicate IP assigned before the allocator cache knew about existing assignments.

Populate the node allocator cache with existing EgressIP status assignments in addEgressNode() BEFORE reconciling unassigned EgressIPs. Also add syncEgressIPAllocator() which is called from syncEgressIPMarkAllocator() to ensure the cache is populated when WatchEgressIP() processes ADD events.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved allocator initialization to pre-populate assignments from status, preventing duplicate EgressIP assignments during restarts and concurrent reconciliations.

* **Tests**
  * Added a regression test that verifies duplicate IPs are not reassigned when multiple EgressIP resources share the same spec during restart (test included repeated block).

### CODE REVIEW

File: go-controller/pkg/clustermanager/egressip_controller.go [hunk 1]
```diff
@@ -1793,9 +1804,36 @@ func generateStatusPatchOp(statusItems []egressipv1.EgressIPStatusItem) jsonPatc
...
+}
+
 // syncEgressIPMarkAllocator iterates over all existing EgressIPs. It builds a mark cache of existing marks stored on each
 // EgressIP annotation or allocates and adds a new mark to an EgressIP if it doesn't exist
 func (eIPC *egressIPClusterController) syncEgressIPMarkAllocator(egressIPs []interface{}) error {
+	// First, pre-populate the allocator cache with existing assignments from EgressIP statuses.
+	// This must happen before individual EgressIP ADD events are processed to prevent
+	// duplicate IP assignments when two EgressIPs have the same IP in their specs but only
+	// one has it assigned in its status.
+	if err := eIPC.syncEgressIPAllocator(egressIPs); err != nil {
```
Comment (pliurh) [inline go-controller/pkg/clustermanager/egressip_controller.go]: Can we combine the iterations for `egressIPs` into the following loop? We can reduce one iteration and use a new helper function to contain the duplicate code used in `addEgressNode`.

**Review by pliurh — APPROVED** [review]

**Comment by cgoncalves** [comment]
Hi @trozet, could you please review this PR? Thanks.

File: go-controller/pkg/clustermanager/egressip_controller.go [hunk 2]
```diff
@@ -745,6 +745,15 @@ func (eIPC *egressIPClusterController) addEgressNode(nodeName string) error {
 	if err != nil {
 		return fmt.Errorf("unable to list EgressIPs, err: %v", err)
 	}
+
+	// Before reconciling unassigned EgressIPs, ensure the allocator cache is populated
+	// with existing assignments from EgressIP statuses. This prevents duplicate IP
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller.go]: @pperiyasamy didn't we speak about something very related to this fix?
did you do the same but on the ovnkube-controller side was it? I remember us talking about this problem that on restart we don't have status cache populated.. but that was before my pto so can't remember all the details...

Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/egressip_controller.go]: @tssurya The ovnkube-controller side fix was about handling EIP failover scenarios properly so that it doesn't lead to any stale SNAT and LRP entries at OVN (i.e. avoid referring to previously assigned node), but this PR is handling EIP configuration issue (two EIP objects having same IP) at cluster manager. both are different issues. The ovnkube-controller assigns active and standby EgressIPs for a pod using EIP name,  so fixing this at cluster manager makes sense.

Comment (tssurya) [reply go-controller/pkg/clustermanager/egressip_controller.go]: yea agreed, just was checking if that was on ovnkube-controller side or CM side.. but yea thanks for jogging my memory

File: go-controller/pkg/clustermanager/egressip_controller.go [hunk 3]
```diff
@@ -745,6 +745,15 @@ func (eIPC *egressIPClusterController) addEgressNode(nodeName string) error {
...
+	// assignments when two EgressIPs have the same IP in their specs but only one has
+	// it assigned in status (e.g., after control-plane restart or during initial sync).
+	for _, egressIP := range egressIPs {
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller.go]: @cgoncalves hmm could you explain why you also need it from addEgressNode? - because IIUC addEgressNode will be called for every node event, its not a one time thing at start up right?
Don't we need to sync only once on startup which is what you do from the sync function?

File: go-controller/pkg/clustermanager/egressip_controller.go [hunk 4]
```diff
@@ -1793,15 +1802,29 @@ func generateStatusPatchOp(statusItems []egressipv1.EgressIPStatusItem) jsonPatc
...
+// during restart when EgressIPs are processed in arbitrary order.
+func (eIPC *egressIPClusterController) ensureAllocatorEgressIPAssignments(egressIP *egressipv1.EgressIP) {
+	if len(egressIP.Status.Items) > 0 {
+		eIPC.addAllocatorEgressIPAssignments(egressIP.Name, egressIP.Status.Items)
+	}
+}
+
 // syncEgressIPMarkAllocator iterates over all existing EgressIPs. It builds a mark cache of existing marks stored on each
 // EgressIP annotation or allocates and adds a new mark to an EgressIP if it doesn't exist
 func (eIPC *egressIPClusterController) syncEgressIPMarkAllocator(egressIPs []interface{}) error {
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller.go:1858]: probably time to rename this to `syncEgressIPs` if we bring generic sync as well into this function.

File: go-controller/pkg/clustermanager/egressip_controller.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/egressip_controller.go]: I think I know why you had to do this.. its cause we start the nodes before we start the egressip watch
```
func (eIPC *egressIPClusterController) Start() error {
	var err error
	// In cluster manager, we only need to watch for egressNodes, egressIPs
	// and cloudPrivateIPConfig
	if eIPC.egressNodeHandler, err = eIPC.WatchEgressNodes(); err != nil {
		return fmt.Errorf("unable to watch egress nodes %w", err)
	}
	if eIPC.egressIPHandler, err = eIPC.WatchEgressIP(); err != nil {
		return err
	}
```

So the right spot for the fix to go in is inside

```
		switch h.objType {
		case factory.EgressIPType:
			syncFunc = h.eIPC.syncEgressIPMarkAllocator
		case factory.EgressNodeType:
			syncFunc = h.eIPC.initEgressNodeReachability
```
initEgressNodeReachability which seems to be the first func we call but we don't have the objects available there but we could just use the eIPC.kube.GetEgressIPs() from there which is what this function seems to be doing?
because otherwise I don't see how the call can help when done after watchnodes...

Comment (cgoncalves) [reply go-controller/pkg/clustermanager/egressip_controller.go]: You're right! This could be just a one-time sync. I moved the EgressIP status-to-allocator-cache population logic to `initEgressNodeReachability`.

File: go-controller/pkg/clustermanager/egressip_controller.go [hunk 5]
```diff
@@ -1825,15 +1844,30 @@ func generateStatusPatchOp(statusItems []egressipv1.EgressIPStatusItem) jsonPatc
...
+	// assignments from EgressIP statuses. The allocator cache must be populated before
+	// individual EgressIP ADD events are processed to prevent duplicate IP assignments
+	// when two EgressIPs have the same IP in their specs but only one has it assigned in
+	// its status.
 	for _, object := range egressIPs {
 		egressIP, ok := object.(*egressipv1.EgressIP)
 		if !ok {
 			return fmt.Errorf("failed to cast %T to *egressipv1.EgressIP", egressIP)
 		}
+		eIPC.ensureAllocatorEgressIPAssignments(egressIP)
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller.go]: so do we still need to call it from syncEgressIP here? if you agree with the logic discussed here: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5760#discussion_r2631115304 - I think all we need to do is call it once from syncNode part? by the time we call syncEIP things cache should be already populated? Since I don't know the full context of the bug would be good if you can test it out and see if you need to call it from here...

Comment (cgoncalves) [reply go-controller/pkg/clustermanager/egressip_controller.go]: The call to `ensureAllocatorEgressIPAssignments` in `syncEgressIPs` is technically redundant but serves as a one-time bulk-sync safety measure to ensure the allocator cache is fully populated during the initial EgressIP informer sync. Since it is idempotent and runs only once at startup, keeping it provides extra robustness against edge cases where a node might not have been ready when addEgressNode first ran.

During testing for the bug this PR aims to fix, I have not ran into an issue. I can drop the call to `ensureAllocatorEgressIPAssignments` if you think that's best, let me know.

Comment (tssurya) [reply go-controller/pkg/clustermanager/egressip_controller.go]: yea I think let's drop the call from syncEgressIPs unless we know we need it (and we can keep this as syncEgressIPMarkAllocator
once you repush that I think we are close to merge (I'll review the UTs again in the next push)

File: go-controller/pkg/clustermanager/egressip_controller_test.go [hunk 6]
```diff
@@ -4218,6 +4218,204 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
...
+					// Both EgressIPs exist at startup - simulating restart scenario
+					&egressipv1.EgressIPList{Items: []egressipv1.EgressIP{eIP1, eIP2}},
+				)
+
+				// Use WatchEgressNodes to properly initialize the allocator cache
+				// (simulating real startup behavior rather than manually setting up cache)
+				_, err := fakeClusterManagerOVN.eIPC.WatchEgressNodes()
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+				_, err = fakeClusterManagerOVN.eIPC.WatchEgressIP()
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller_test.go]: ~then I think we can remove this part as well as its triggered via WatchEgressNodes() already~
ah I guess we need to watchIP for the status update stuff to kick in nvm

File: go-controller/pkg/clustermanager/egressip_controller_test.go [hunk 7]
```diff
@@ -4218,6 +4218,204 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
...
+
+				// eIP1 should keep its assignment (the IP was already assigned)
+				gomega.Eventually(getEgressIPStatusLen("egressip-1")).Should(gomega.Equal(1))
+				egressIPs1, nodes1 := getEgressIPStatus("egressip-1")
+				gomega.Expect(nodes1[0]).To(gomega.Equal(node1Name))
+				gomega.Expect(egressIPs1[0]).To(gomega.Equal(duplicateIP))
+
+				// eIP2 should NOT get the duplicate IP assigned - it should remain unassigned
+				// because syncEgressIPs pre-populated the cache with eIP1's assignment
+				gomega.Eventually(getEgressIPStatusLen("egressip-2")).Should(gomega.Equal(0))
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller_test.go]: nice test!

File: go-controller/pkg/clustermanager/egressip_controller_test.go [hunk 8]
```diff
@@ -4218,6 +4218,204 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
...
+				return nil
+			}
+
+			err := app.Run([]string{app.Name})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		})
+
+		// This test validates that with multiple nodes, a duplicate IP is not assigned
+		// to a different node during restart. The IP should stay with the original node.
+		ginkgo.It("should not assign duplicate IP to different node during restart", func() {
```
Comment (tssurya) [inline go-controller/pkg/clustermanager/egressip_controller_test.go]: hmm I don't get the difference between the previous UT and this one... shall we just combine them into having one test?

Comment (cgoncalves) [reply go-controller/pkg/clustermanager/egressip_controller_test.go]: You're right. Both tests validate duplicate IP prevention during a restart scenario, but they differ in the number of nodes and the specific edge case being tested.

- Test 1: validates that the second EgressIP doesn't get assigned the duplicate IP when there's only one available node.
- Test 2: validates that the duplicate IP won't be assigned to a different node (node2), even when another eligible node is available

Test 2 is more comprehensive. It ensures the duplicate IP detection works globally across all nodes, not just preventing double-assignment on a single node. Without Test 2, a bug could exist where the system correctly prevents assigning to node1 but incorrectly assigns the duplicate to node2. So, test 2 is a superset of the first test.

Updated the code. Thanks fo the great review, Surya!

**Comment by tssurya** [comment]
```
Summarizing 2 Failures:
  [FAIL] Services of type NodePort [It] should work on secondary node interfaces for ETP=local and ETP=cluster when backend pods are also served by EgressIP [Feature:Service]
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/service.go:1421
  [FAIL] e2e ingress traffic validation Validating ingress traffic [It] Should be allowed by nodeport services
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1335
```
https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/21667639243/job/62468109789?pr=5760
I've never seen these failures before.. @cgoncalves see if you can double check on these but they are on ic-disabled lane so I will retest it meanwhile

**Comment by tssurya** [comment]
```
2026-02-04T10:56:13.7579599Z   [1mAttempt #1 [38;5;9mFailed[0m[1m.  Retrying ↺[0m [38;5;243m@ 02/04/26 10:56:13.756[0m
2026-02-04T10:56:13.7581332Z 
2026-02-04T10:56:13.7582226Z   [1mSTEP:[0m Creating a kubernetes client [38;5;243m@ 02/04/26 10:56:13.756[0m
2026-02-04T10:56:13.7583054Z   I0204 10:56:13.756932 63460 util.go:454] >>> kubeConfig: /home/runner/ovn.conf
2026-02-04T10:56:13.7583894Z   [1mSTEP:[0m Building a namespace api object, basename nodeport-ingress-test [38;5;243m@ 02/04/26 10:56:13.757[0m
2026-02-04T10:56:13.7692282Z   [1mSTEP:[0m Waiting for a default service account to be provisioned in namespace [38;5;243m@ 02/04/26 10:56:13.768[0m
2026-02-04T10:56:13.7733256Z   [1mSTEP:[0m Waiting for kube-root-ca.crt to be provisioned in namespace [38;5;243m@ 02/04/26 10:56:13.773[0m
2026-02-04T10:56:13.7782882Z   [1mSTEP:[0m Creating the endpoints pod, one for each worker [38;5;243m@ 02/04/26 10:56:13.778[0m
2026-02-04T10:56:25.8334777Z   [1mSTEP:[0m Creating an external container to send the traffic from [38;5;243m@ 02/04/26 10:56:25.833[0m
2026-02-04T10:56:25.8789390Z creating container with command: "run -itd --privileged --name e2e-ingress --network kind --hostname e2e-ingress registry.k8s.io/e2e-test-images/agnhost:2.56 netexec --http-port=12006"
2026-02-04T10:56:26.3633458Z   [1mSTEP:[0m Creating the nodeport service [38;5;243m@ 02/04/26 10:56:26.363[0m
2026-02-04T10:56:26.3738905Z   [1mSTEP:[0m Waiting for the endpoints to pop up [38;5;243m@ 02/04/26 10:56:26.372[0m
2026-02-04T10:56:27.3724264Z   I0204 10:56:27.372192 63460 util.go:419] Waiting for amount of service:nodeportsvc endpoints to be 3
2026-02-04T10:56:27.3750417Z   [1mSTEP:[0m Hitting the nodeport on ovn-control-plane and reaching all the endpoints http [38;5;243m@ 02/04/26 10:56:27.374[0m
2026-02-04T10:56:32.4329964Z   I0204 10:56:32.432667 63460 util.go:333] FAILED Command was [curl -g -q -s http://localhost:12006/dial?request=hostname&protocol=http&host=fc00:f853:ccd:e793::3&port=30601&tries=1]
2026-02-04T10:56:32.4332160Z   I0204 10:56:32.432716 63460 util.go:334] FAILED Response was {"errors":["Get \"http://[fc00:f853:ccd:e793::3]:30601/hostname\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"]}
```
aaaand
```
2026-02-04T10:52:41.5929402Z [0me2e ingress traffic validation [38;5;243mValidating ingress traffic [0m[1mShould be allowed by nodeport services[0m
2026-02-04T10:52:41.5931806Z [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1292[0m
2026-02-04T10:52:41.5933108Z   [1mSTEP:[0m Creating a kubernetes client [38;5;243m@ 02/04/26 10:52:41.591[0m
2026-02-04T10:52:41.5934353Z   I0204 10:52:41.591298 63460 util.go:454] >>> kubeConfig: /home/runner/ovn.conf
2026-02-04T10:52:41.5935838Z   [1mSTEP:[0m Building a namespace api object, basename nodeport-ingress-test [38;5;243m@ 02/04/26 10:52:41.592[0m
2026-02-04T10:52:41.6066127Z   [1mSTEP:[0m Waiting for a default service account to be provisioned in namespace [38;5;243m@ 02/04/26 10:52:41.605[0m
2026-02-04T10:52:41.6116036Z   [1mSTEP:[0m Waiting for kube-root-ca.crt to be provisioned in namespace [38;5;243m@ 02/04/26 10:52:41.61[0m
2026-02-04T10:52:41.6164435Z   [1mSTEP:[0m Creating the endpoints pod, one for each worker [38;5;243m@ 02/04/26 10:52:41.616[0m
2026-02-04T10:52:53.6620539Z   [1mSTEP:[0m Creating an external container to send the traffic from [38;5;243m@ 02/04/26 10:52:53.661[0m
2026-02-04T10:52:53.7040462Z creating container with command: "run -itd --privileged --name e2e-ingress --network kind --hostname e2e-ingress registry.k8s.io/e2e-test-images/agnhost:2.56 netexec --http-port=12005"
2026-02-04T10:52:54.1794402Z   [1mSTEP:[0m Creating the nodeport service [38;5;243m@ 02/04/26 10:52:54.179[0m
2026-02-04T10:52:54.1964474Z   [1mSTEP:[0m Waiting for the endpoints to pop up [38;5;243m@ 02/04/26 10:52:54.195[0m
2026-02-04T10:52:55.1968679Z   I0204 10:52:55.196623 63460 util.go:419] Waiting for amount of service:nodeportsvc endpoints to be 3
2026-02-04T10:52:55.1992445Z   [1mSTEP:[0m Hitting the nodeport on ovn-control-plane and reaching all the endpoints http [38;5;243m@ 02/04/26 10:52:55.199[0m
2026-02-04T10:53:00.2680589Z   I0204 10:53:00.265041 63460 util.go:333] FAILED Command was [curl -g -q -s http://localhost:12005/dial?request=hostname&protocol=http&host=fc00:f853:ccd:e793::3&port=31535&tries=1]
2026-02-04T10:53:00.2693472Z   I0204 10:53:00.265078 63460 util.go:334] FAILED Response was {"errors":["Get \"http://[fc00:f853:ccd:e793::3]:31535/hostname\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"]}
2026-02-04T10:53:05.3313849Z   I0204 10:53:05.331135 63460 util.go:333] FAILED Command was [curl -g -q -s http://localhost:12005/dial?request=hostname&protocol=http&host=fc00:f853:ccd:e793::3&port=31535&tries=1]
```
and
```
2026-02-04T10:51:41.9877460Z   I0204 10:51:41.987066 63460 service.go:1314] skipping IP fc00:f853:ccd:e793::4 because it is not within networks subnet 2001:db8:abcd:1234::/64
2026-02-04T10:51:41.9878395Z   [1mSTEP:[0m Adding additional IP addresses to each node [38;5;243m@ 02/04/26 10:51:41.987[0m
2026-02-04T10:51:43.1269045Z   [1mSTEP:[0m Creating NodePort services [38;5;243m@ 02/04/26 10:51:43.126[0m
2026-02-04T10:51:43.2194953Z   [1mSTEP:[0m Waiting for the endpoints to pop up [38;5;243m@ 02/04/26 10:51:43.218[0m
2026-02-04T10:51:44.2200910Z   I0204 10:51:44.219650 63460 util.go:419] Waiting for amount of service:etp-local-svc endpoints to be 3
2026-02-04T10:51:45.2228423Z   I0204 10:51:45.222477 63460 util.go:419] Waiting for amount of service:etp-cluster-svc endpoints to be 3
2026-02-04T10:51:45.2253808Z   [1mSTEP:[0m Checking connectivity to the external container from egressIP pod ovn-control-plane-ep and verify that the source IP is the secondary NIC egress IP [38;5;243m@ 02/04/26 10:51:45.225[0m
2026-02-04T10:51:45.2256175Z   I0204 10:51:45.225107 63460 service.go:1395] Destination IPs for external container are ip=2001:db8:abcd:1234::5
2026-02-04T10:51:45.2258270Z   I0204 10:51:45.225186 63460 builder.go:130] Running '/usr/local/bin/kubectl --server=https://[::1]:39631 --kubeconfig=/home/runner/ovn.conf --namespace=services-6268 exec ovn-control-plane-ep -- curl --connect-timeout 2 [2001:db8:abcd:1234::5]:12003/clientip'
2026-02-04T10:51:45.3218371Z   I0204 10:51:45.321425 63460 builder.go:156] stderr: "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r100    33  100    33    0     0   9060      0 --:--:-- --:--:-- --:--:-- 11000\n"
2026-02-04T10:51:45.3219820Z   I0204 10:51:45.321461 63460 builder.go:157] stdout: "[2001:db8:abcd:1234:c001::]:60746"
2026-02-04T10:51:45.3221214Z   [1mSTEP:[0m Hitting service etp-local-svc on 2001:db8:abcd:1234::140 via http [38;5;243m@ 02/04/26 10:51:45.321[0m
2026-02-04T10:51:50.3803624Z   I0204 10:51:50.380043 63460 util.go:333] FAILED Command was [curl -g -q -s http://localhost:12003/dial?request=hostname&protocol=http&host=2001:db8:abcd:1234::140&port=32718&tries=1]
2026-02-04T10:51:50.3805804Z   I0204 10:51:50.380081 63460 util.go:334] FAILED Response was {"errors":["Get \"http://[2001:db8:abcd:1234::140]:32718/hostname\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"]}
2026-02-04T10:51:56.4436866Z   I0204 10:51:56.443299 63460 util.go:333] FAILED Command was [curl -g -q -s http://localhost:12003/dial?request=hostname&protocol=http&host=2001:db8:abcd:1234::140&port=32718&tries=1]
```
hmm I'm going to retest to see if its still there but this is a flake I've not seen before on other PRs
